lib/registry.go

```diff
@@ -30,12 +34,65 @@ func NewRegistryRequests(r repo.Repo, cli *rpc.Client) *RegistryRequests {
 	}
 }
 
+// SetQriNode assigns the unexported qri node pointer
+func (r *RegistryRequests) SetQriNode(node *p2p.QriNode) {
+	r.node = node
+}
+
+// PublishParams encapsulates arguments to the publish method
+type PublishParams struct {
+	Ref repo.DatasetRef
+	Pin bool
+}
+
 // Publish a dataset to a registry
-func (r *RegistryRequests) Publish(ref *repo.DatasetRef, done *bool) error {
+func (r *RegistryRequests) Publish(p *PublishParams, done *bool) (err error) {
 	if r.cli != nil {
-		return r.cli.Call("RegistryRequests.Publish", ref, done)
+		return r.cli.Call("RegistryRequests.Publish", p, done)
+	}
+
+	ref := p.Ref
+
+	if p.Pin {
+		log.Info("pinning dataset...")
+		node := r.node
+
+		if node == nil {
+			// if we don't have an online node, create one and connect
+			node, err = p2p.NewQriNode(r.repo.Repo, func(c *config.P2P) {
+				*c = *Config.P2P
+				c.Enabled = true
+			})
+			if err != nil {
+				return err
+			}
+
+			if err := node.StartOnlineServices(func(string) {}); err != nil {
```

Comment:
Extract this common code and combine with the next condition, change from "else if" to "if".